src/main/java/org/jitsi/jibri/service/impl/FileRecordingJibriService.kt

```diff
@@ -0,0 +1,158 @@
+
+/**
+ * [FileRecordingJibriService] is the [JibriService] responsible for joining
+ * a web call, capturing its audio and video, and writing that audio and video
+ * to a file to be replayed later.
+ */
+class FileRecordingJibriService(private val recordingOptions: RecordingOptions) : JibriService() {
+    /**
+     * The [Logger] for this class
+     */
+    private val logger = Logger.getLogger(this::class.qualifiedName)
+    /**
+     * Used for the selenium interaction
+     */
+    private val jibriSelenium = JibriSelenium(
+        JibriSeleniumOptions(recordingOptions.callParams),
+        Executors.newSingleThreadScheduledExecutor(NameableThreadFactory("JibriSelenium"))
+    )
+    /**
+     * The [FfmpegCapturer] that will be used to capture media from the call and write it to a file
+     */
+    private val capturer = FfmpegCapturer()
+    /**
+     * The [Sink] this class will use to model the file on the filesystem
+     */
+    private var sink: Sink
+    /**
+     * If ffmpeg dies for some reason, we want to restart it.  This [ScheduledExecutorService]
+     * will run the process monitor in a separate thread so it can check that it's running on its own
+     */
+    private val executor: ScheduledExecutorService =
+        Executors.newSingleThreadScheduledExecutor(NameableThreadFactory("FileRecordingJibriService"))
+    /**
+     * The handle to the scheduled process monitor task, which we use to
+     * cancel the task
+     */
+    private var processMonitorTask: ScheduledFuture<*>? = null
+
+    init {
+        sink = FileSink(recordingOptions.recordingDirectory, recordingOptions.callParams.callUrlInfo.callName)
+        jibriSelenium.addStatusHandler {
+            publishStatus(it)
+        }
+    }
+
+    /**
+     * @see [JibriService.start]
+     */
+    override fun start(): Boolean {
+        if (!jibriSelenium.joinCall(recordingOptions.callParams.callUrlInfo.callName)) {
+            logger.error("Selenium failed to join the call")
+            stop()
+            return false
+        }
+        capturer.start(sink)
+        var numRestarts = 0
+        val processMonitor = ProcessMonitor(capturer) { exitCode ->
+            if (exitCode != null) {
+                logger.error("Capturer process is no longer healthy.  It exited with code $exitCode")
+            } else {
+                logger.error("Capturer process is no longer healthy but it is still running, stopping it now")
+                capturer.stop()
+            }
+            if (numRestarts == FFMPEG_RESTART_ATTEMPTS) {
+                logger.error("Giving up on restarting the capturer")
+                publishStatus(JibriServiceStatus.ERROR)
+                stop()
+            } else {
+                numRestarts++
+                // Re-create the sink here because we want a new filename
+                sink = FileSink(recordingOptions.recordingDirectory, recordingOptions.callParams.callUrlInfo.callName)
+                capturer.start(sink)
```

Comment:
Does this need some kind of backoff or wait time or is immediate retry sufficient?

Reply:
nah no backoff needed here